### PR TITLE
pessimistic-txn: implement pessimistic rollback

### PR DESCRIPTION
**Pull request by youjiali1995**
## What have you changed? (mandatory)
* Add `pessimistic_rollback` API to delete any pessimistic locks with small for_update_ts belongs to  the transaction. 
* `acquire_pessimistic_lock` will overwrite own pessimistic locks with small for_update_ts.
* Now key hash is generated from raw key, which consumes more CPU.

## What are the type of the changes? (mandatory)
- New feature (change which adds functionality)

## How has this PR been tested? (mandatory)
unit tests and manual tests.

## Does this PR affect documentation (docs) or release note? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Refer to a related PR or issue link (optional)
https://github.com/pingcap/tidb/pull/10654

### CODE REVIEW

**Comment by youjiali1995** [comment]
@coocood @zhangjinpeng1987 @AndreMouche PTAL

File: src/storage/lock_manager/util.rs [hunk 1]
```diff
@@ -2,32 +2,45 @@
...
+    farmhash::fingerprint64(&key.to_raw().unwrap())
 }
 
 pub fn gen_key_hashes(keys: &[Key]) -> Vec<u64> {
     keys.iter().map(|key| gen_key_hash(key)).collect()
 }
 
+pub fn extract_raw_key_from_process_result(pr: &ProcessResult) -> &[u8] {
+    match pr {
+        ProcessResult::MultiRes { results } => {
```
Comment (coocood) [inline src/storage/lock_manager/util.rs]: I think we don't need to panic if failed to extract the key, as it will only be used for logging.

Comment (youjiali1995) [reply src/storage/lock_manager/util.rs]: It never occurs. If panicked, it's a bug in implementation.

**Comment by coocood** [comment]
LGTM

File: src/storage/lock_manager/util.rs [hunk 2]
```diff
@@ -2,32 +2,45 @@
...
+        _ => panic!("unexpected mvcc error"),
     }
 }
 
+// TiDB uses the same hash algorithm.
 pub fn gen_key_hash(key: &Key) -> u64 {
-    let mut s = DefaultHasher::new();
-    key.hash(&mut s);
-    s.finish()
+    farmhash::fingerprint64(&key.to_raw().unwrap())
```
Comment (zhangjinpeng1987) [inline src/storage/lock_manager/util.rs]: Please give some benchmark for farmhash.

Comment (youjiali1995) [reply src/storage/lock_manager/util.rs]:
```
running 2 tests
test bench_farmhash ... bench:          19 ns/iter (+/- 0)
test bench_std_hash ... bench:          35 ns/iter (+/- 1)
```

TiDB will check `deadlock_key_hash` to judge is it a single statement deadlock. So TiKV should use the same hash algorithm with TiDB.

**Pull request merged**